segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,382 @@
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+
+import static java.nio.file.attribute.PosixFilePermission.OWNER_WRITE;
+
+/**
+ * Storage adapter for file system based Tier 2.
+ *
+ * Each segment is represented as a single file on the underlying storage. As the data in Tier 2 is not modified
+ * once written, any attempt to re-write data with the same file offset does not cause any form of inconsistency
+ * as the bytes are the same and they are in the same position.
+ */
+
+@Slf4j
+public class FSStorage implements Storage {
+
+    //region members
+
+    private final FSStorageConfig config;
+    private final ExecutorService executor;
+
+    //endregion
+
+    //region constructor
+
+    /**
+     * Creates a new instance of the FSStorage class.
+     *
+     * @param config   The configuration to use.
+     * @param executor The executor to use for running async operations.
+     */
+    public FSStorage(FSStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+
+    /**
+     * Initialize is a no op here as we do not need a locking mechanism in case of file system write.
+     * @param containerEpoch The Container Epoch to initialize with (ignored here).
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        return CompletableFuture.supplyAsync( () -> syncOpenRead(streamSegmentName),
+                executor);
+    }
+
+
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
+            length, Duration timeout) {
+        return CompletableFuture.supplyAsync(() -> syncRead(handle, offset, buffer, bufferOffset, length, timeout),
+                executor);
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        return CompletableFuture.supplyAsync(() -> syncGetStreamSegmentInfo(streamSegmentName, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        return CompletableFuture.supplyAsync(() -> syncExists(streamSegmentName, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return CompletableFuture.supplyAsync(() -> {
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            if (!Files.exists(path)) {
```

Comment:
The reason for this implementation is the definition of openRead and openWrite APIs. If this is not handled here, these methods return successfully and failure is observed when these handles are used.
But these methods are expected to check and throw the exceptions (Please see `ReadOnlyStorage.java`)